File: SHLasso-KMFM-paper/libs/SHIM_4way_CB.R
source("Updates.R")

# SHIM CLASS for 4-way interactions and adaptive estimation of main effects
SHIM_4way <- function(X,
                      y,
                      beta_init,
                      gamma_init,
                      delta_init,
                      tau_init,
                      l1 = 21,
                      l2 = 14,
                      l3 = 2,
                      l4 = 3)
  
{
  self = list()
  
  self$beta_hat <- beta_init
  self$gamma_hat <- gamma_init
  self$delta_hat <- delta_init
  self$tau_hat <- tau_init
  
  self$means_X <- colMeans(X)
  self$stds_X <- apply(X, 2, sd)
  self$mean_y <- mean(y)
  self$scale = scale
  
  self$l1 = l1
  self$l2 = l2
  self$l3 = l3
  self$l4 = l4
  
  
  fit <- function(X,
                  y,
                  lambda_beta,
                  lambda_gamma,
                  lambda_delta,
                  lambda_tau,
                  w_beta = NULL,
                  w_gamma = 1,
                  w_delta = 1,
                  w_tau = 1,
                  tol = 5e-3,
                  max_iter = 10,
                  compute_Q = Q_bern,
                  intercept = 0,
                  use_intercept = TRUE) #Take care! Usually intercept has to be used!
  {
    # STEP 0 (INITIALIZE AND UPDATE THE INTERCEPT)
    beta_hat <- self$beta_hat
    gamma_hat <- self$gamma_hat
    delta_hat <- self$delta_hat
    tau_hat <- self$tau_hat
    Q_old <- 1e100
    beta_2way <- get_beta_vec_2way4(
      beta = beta_hat,
      l1 = l1,
      l2 = l2,
      l3 = l3,
      l4 = l4,
      gamma = gamma_hat,
      only_beta = FALSE
    )
    beta_3way <- get_beta_vec_3way4(
      beta_2way = beta_2way,
      l1 = l1,
      l2 = l2,
      l3 = l3,
      l4 = l4,
      delta = delta_hat,
      only_beta = FALSE
    )
    beta_4way <- get_beta_vec_4way4(
      beta_3way = beta_3way,
      l1 = l1,
      l2 = l2,
      l3 = l3,
      l4 = l4,
      tau = tau_hat,
      only_beta = FALSE
    )
    beta_all <- c(beta_hat, beta_2way, beta_3way, beta_4way)
    
    # Do the updates until convergence or maximum number of iterations is reached
    for (i in c(1:max_iter))
    { 
      cat(" Start iteration: ", i, " ; ")
      
      if (use_intercept == TRUE)
      {
        # Update the intercept
        intercept <- update_intercept(
          X = X,
          y = y,
          beta_all = beta_all,
          intercept_old = intercept
        ) 
      }
      
      # STEP 1 (UPDATE TAU HAT)
      tau_hat <- update_tau(
        X = X,
        y = y,
        beta_hat = beta_hat,
        gamma_hat = gamma_hat,
        delta_hat = delta_hat,
        tau_hat = tau_hat,
        lambda_tau = lambda_tau,
        l1 = self$l1,
        l2 = self$l2,
        l3 = self$l3,
        l4 = self$l4,
        intercept = intercept
      )
      
      # STEP 2 (UPDATE DELTA HAT)
      delta_hat <- update_delta(
        X = X,
        y = y,
        beta_hat = beta_hat,
        gamma_hat = gamma_hat,
        delta_hat = delta_hat,
        tau_hat = tau_hat,
        lambda_delta = lambda_delta,
        l1 = self$l1,
        l2 = self$l2,
        l3 = self$l3,
        l4 = self$l4,
        intercept = intercept
      )
      
      # STEP 3 (UPDATE GAMMA HAT)
      gamma_hat <- update_gamma(
        X = X,
        y = y,
        beta_hat = beta_hat,
        gamma_hat = gamma_hat,
        delta_hat = delta_hat,
        tau_hat = tau_hat,
        lambda_gamma = lambda_gamma,
        l1 = self$l1,
        l2 = self$l2,
        l3 = self$l3,
        l4 = self$l4,
        intercept = intercept
      )
      
      # STEP 4 (UPDATE BETA HAT)
      beta_hat <- update_beta(
        X = X,
        y = y,
        beta_hat = beta_hat,
        gamma_hat = gamma_hat,
        delta_hat = delta_hat,
        tau_hat = tau_hat,
        lambda_beta = lambda_beta,
        intercept = intercept,
        l1 = self$l1,
        l2 = self$l2,
        l3 = self$l3,
        l4 = self$l4,
        w = w_beta
      )
      
      
      # Keep track of the updates
      beta_2way <- get_beta_vec_2way4(beta = beta_hat, l1=l1, l2=l2, l3=l3, l4=l4, gamma=gamma_hat, only_beta = FALSE) 
      beta_3way <- get_beta_vec_3way4(beta_2way = beta_2way, l1=l1, l2=l2, l3=l3, l4=l4, delta=delta_hat, only_beta = FALSE) 
      beta_4way <- get_beta_vec_4way4(beta_3way = beta_3way, l1=l1, l2=l2, l3=l3, l4=l4, tau=tau_hat, only_beta = FALSE) 
      beta_all<-c(beta_hat, beta_2way, beta_3way, beta_4way)
      
      
      # STEP 5 (COMPUTE RELATIVE DIFFERENCE AND DECIDE IF THE ALGORITHM STOPS)
      Q_new <- compute_Q(
        X = X,
        y = y,
        beta = beta_hat,
        gamma_vec = gamma_hat,
        delta_vec = delta_hat,
        tau_vec = tau_hat,
        lambda_beta = lambda_beta,
        lambda_gamma = lambda_gamma,
        lambda_delta = lambda_delta,
        lambda_tau = lambda_tau,
        w_beta = w_beta,
        w_gamma = w_gamma,
        w_delta = w_delta,
        w_tau = w_tau,
        l1 = self$l1,
        l2 = self$l2,
        l3 = self$l3,
        l4 = self$l4,
        intercept = intercept
      )
      if (Q_new == Q_old)
        # If equal then stop. This check ensures that the algorithm stops when Q_new=Q_old=0.
      {
        self$beta_hat <- beta_hat
        self$gamma_hat <- gamma_hat
        self$delta_hat <- delta_hat
        self$tau_hat <- tau_hat
        beta_2way <- get_beta_vec_2way4(
          beta = beta_hat,
          l1 = l1,
          l2 = l2,
          l3 = l3,
          l4 = l4,
          gamma = gamma_hat,
          only_beta = FALSE
        )
        beta_3way <- get_beta_vec_3way4(
          beta_2way = beta_2way,
          l1 = l1,
          l2 = l2,
          l3 = l3,
          l4 = l4,
          delta = delta_hat,
          only_beta = FALSE
        )
        beta_4way <- get_beta_vec_4way4(
          beta_3way = beta_3way,
          l1 = l1,
          l2 = l2,
          l3 = l3,
          l4 = l4,
          tau = tau_hat,
          only_beta = FALSE
        )
        beta_all <- c(beta_hat, beta_2way, beta_3way, beta_4way)
        self$intercept <- intercept
        print("The model has converged.")
        return (
          list(
            "beta_hat" = self$beta_hat,
            "gamma_hat" = self$gamma_hat ,
            "delta_hat" = self$delta_hat,
            "tau_hat" = tau_hat,
            "beta_all" = beta_all,
            "intercept" = self$intercept
          )
        )
      }
      # If small enough then stop
      rel_dif <- compute_rel_dif(Q_old = Q_old, Q_new = Q_new)
      cat("  Relative difference is ", rel_dif)
      if (abs(rel_dif) <= tol) {
        self$beta_hat <- beta_hat
        self$gamma_hat <- gamma_hat
        self$delta_hat <- delta_hat
        self$tau_hat <- tau_hat
        beta_2way <- get_beta_vec_2way4(
          beta = beta_hat,
          l1 = l1,
          l2 = l2,
          l3 = l3,
          l4 = l4,
          gamma = gamma_hat,
          only_beta = FALSE
        )
        beta_3way <- get_beta_vec_3way4(
          beta_2way = beta_2way,
          l1 = l1,
          l2 = l2,
          l3 = l3,
          l4 = l4,
          delta = delta_hat,
          only_beta = FALSE
        )
        beta_4way <- get_beta_vec_4way4(
          beta_3way = beta_3way,
          l1 = l1,
          l2 = l2,
          l3 = l3,
          l4 = l4,
          tau = tau_hat,
          only_beta = FALSE
        )
        beta_all <- c(beta_hat, beta_2way, beta_3way, beta_4way)
        self$intercept <- intercept
        print("The model has converged.")
        return (
          list(
            "beta_hat" = self$beta_hat,
            "gamma_hat" = self$gamma_hat ,
            "delta_hat" = self$delta_hat,
            "tau_hat" = tau_hat,
            "beta_all" = beta_all,
            "intercept" = self$intercept
          )
        )
      }
      Q_old <- Q_new #Update Q_old
    }
    cat(
      "It has not converged. The max number of iterations can be increased."
    )
    self$beta_hat <- beta_hat
    self$gamma_hat <- gamma_hat
    self$delta_hat <- delta_hat
    self$tau_hat <- tau_hat
    beta_2way <- get_beta_vec_2way4(
      beta = beta_hat,
      l1 = l1,
      l2 = l2,
      l3 = l3,
      l4 = l4,
      gamma = gamma_hat,
      only_beta = FALSE
    )
    beta_3way <- get_beta_vec_3way4(
      beta_2way = beta_2way,
      l1 = l1,
      l2 = l2,
      l3 = l3,
      l4 = l4,
      delta = delta_hat,
      only_beta = FALSE
    )
    beta_4way <- get_beta_vec_4way4(
      beta_3way = beta_3way,
      l1 = l1,
      l2 = l2,
      l3 = l3,
      l4 = l4,
      tau = tau_hat,
      only_beta = FALSE
    )
    beta_all <- c(beta_hat, beta_2way, beta_3way, beta_4way)
    self$intercept <- intercept
    return (
      list(
        "beta_hat" = self$beta_hat,
        "gamma_hat" = self$gamma_hat ,
        "delta_hat" = self$delta_hat,
        "tau_hat" = tau_hat,
        "beta_all" = beta_all,
        "intercept" = self$intercept
      )
    )
  }
  
  predict <- function(self, X_new)
  {
    beta_all <- self$beta_all
    v <-  X_new %*% beta_all + self$intercept
    y_pred <- kappa1(v)
    return(y_pred)
  }
  
  R2_score <- function(self, X_new, y_true, verbose = TRUE)
  {
    y_pred <- predict(self, X_new, scale = scale)
    if (verbose == TRUE)
    {
      cat("R2 score is", r2(y_true, y_pred), "\n")
      plot(array(y_pred), array(y_true),
           xlab = "Predicted Values",
           ylab = "True Values",
           main = "True vs Predicted Values")
      
    }
    
    return(r2(y_true, y_pred))
  }
  
  return(list(
    fit = fit,
    predict = predict,
    R2_score = R2_score,
    self = self
  ))
  
}

File: SHLasso-KMFM-paper/KMFM_main_paper/Strong_hierarchical_lasso/libs_4way_cb/SHIM_4way_CB.py
#import sys
#import os

# Add ../libs to the Python path
#sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "libs")))

from .Updates import *


class SHIM_4way_CB:
    def __init__(self, X, y, beta_init, gamma_init, delta_init, tau_init, 
                 l1=21, l2=14, l3=2, l4=3):



        range_main, range_theta, range_psi, range_phi=get_ranges4(l1,l2,l3,l4)
        
        if beta_init is None:
            beta_init = np.zeros(len(range_main))
        if gamma_init is None:
            gamma_init = np.zeros(len(range_theta))
        if delta_init is None:
            delta_init=np.zeros(len(range_psi)) 
        if tau_init is None:
            tau_init = np.zeros(len(range_phi))

        # store copies to avoid aliasing with caller
        self.beta_hat  = np.array(beta_init,  dtype=float).copy()
        self.gamma_hat = np.array(gamma_init, dtype=float).copy()
        self.delta_hat = np.array(delta_init, dtype=float).copy()
        self.tau_hat   = np.array(tau_init,   dtype=float).copy()

        self.means_X = np.mean(X, axis=0)
        self.stds_X  = np.std(X, axis=0, ddof=1)
        self.mean_y  = float(np.mean(y))
        self.scale   = None

        self.l1, self.l2, self.l3, self.l4 = l1, l2, l3, l4

        self.intercept = 0.0
        self.beta_all  = None

    def fit(self, X, y,
            lambda_beta, lambda_gamma, lambda_delta, lambda_tau,
            w_beta=1, w_gamma=1, w_delta=1, w_tau=1,
            tol=5e-3, max_iter=10, compute_Q=Q_bern,
            intercept=0.0, use_intercept=True):
        #Accept none as init for w
        if w_beta is None:
            w_beta=1
        if w_gamma is None:
            w_gamma=1
        if w_delta is None:
            w_delta=1
        if w_tau is None:
            w_tau=1

        # local working copies
        beta_hat  = self.beta_hat.copy()
        gamma_hat = self.gamma_hat.copy()
        delta_hat = self.delta_hat.copy()
        tau_hat   = self.tau_hat.copy()

        # ensure the intercept is a plain Python float (break possible 0-d np arrays)
        intercept_loc = float(intercept)

        # build initial beta_all
        beta_2way = get_beta_vec_2way4(beta=beta_hat,  l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                       gamma=gamma_hat, only_beta=False)
        beta_3way = get_beta_vec_3way4(beta_2way=beta_2way, l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                       delta=delta_hat, only_beta=False)
        beta_4way = get_beta_vec_4way4(beta_3way=beta_3way, l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                       tau=tau_hat, only_beta=False)
        beta_all = np.concatenate([beta_hat, beta_2way, beta_3way, beta_4way])

        # compute a real Q_old for a meaningful baseline
        Q_old = compute_Q(X=X, y=y, beta=beta_hat,
                          gamma_vec=gamma_hat, delta_vec=delta_hat, tau_vec=tau_hat,
                          lambda_beta=lambda_beta, lambda_gamma=lambda_gamma,
                          lambda_delta=lambda_delta, lambda_tau=lambda_tau,
                          w_beta=w_beta, w_gamma=w_gamma, w_delta=w_delta, w_tau=w_tau,
                          l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                          intercept=intercept_loc)
        #print("beta_hat: ",beta_hat)
        for it in range(1, max_iter + 1):
            print(f" Start iteration: {it};")

            if use_intercept:
                #print("intercept before update:", intercept_loc)
                intercept_loc = float(
                    update_intercept(X=X, y=y, beta_all=beta_all, intercept_old=intercept_loc)
                )
                #print("intercept after  update:", intercept_loc)

            # -- Update beta
            beta_hat = update_beta(X=X, y=y, beta_hat=beta_hat, gamma_hat=gamma_hat,
                                delta_hat=delta_hat, tau_hat=tau_hat, lambda_beta=lambda_beta,
                                intercept=intercept_loc, l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4, w=w_beta)


            # -- Update gamma, delta, tau
            gamma_hat = update_gamma(X=X, y=y, beta_hat=beta_hat, gamma_hat=gamma_hat,
                                    delta_hat=delta_hat, tau_hat=tau_hat, lambda_gamma=lambda_gamma,
                                    l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                    intercept=intercept_loc)

            delta_hat = update_delta(X=X, y=y, beta_hat=beta_hat, gamma_hat=gamma_hat,
                                    delta_hat=delta_hat, tau_hat=tau_hat, lambda_delta=lambda_delta,
                                    l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                    intercept=intercept_loc)

            tau_hat = update_tau(X=X, y=y, beta_hat=beta_hat, gamma_hat=gamma_hat,
                                delta_hat=delta_hat, tau_hat=tau_hat, lambda_tau=lambda_tau,
                                l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,intercept=intercept_loc)

            # rebuild beta_all
            beta_2way = get_beta_vec_2way4(beta=beta_hat,  l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                           gamma=gamma_hat, only_beta=False)
            beta_3way = get_beta_vec_3way4(beta_2way=beta_2way, l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                           delta=delta_hat, only_beta=False)
            beta_4way = get_beta_vec_4way4(beta_3way=beta_3way, l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                                           tau=tau_hat, only_beta=False)
            beta_all = np.concatenate([beta_hat, beta_2way, beta_3way, beta_4way])

            # evaluate objective
            Q_new = compute_Q(X=X, y=y, beta=beta_hat,
                              gamma_vec=gamma_hat, delta_vec=delta_hat, tau_vec=tau_hat,
                              lambda_beta=lambda_beta, lambda_gamma=lambda_gamma,
                              lambda_delta=lambda_delta, lambda_tau=lambda_tau,
                              w_beta=w_beta, w_gamma=w_gamma, w_delta=w_delta, w_tau=w_tau,
                              l1=self.l1, l2=self.l2, l3=self.l3, l4=self.l4,
                              intercept=intercept_loc)

            if Q_new == Q_old:
                self.beta_hat, self.gamma_hat, self.delta_hat, self.tau_hat = beta_hat, gamma_hat, delta_hat, tau_hat
                self.beta_all = beta_all
                self.intercept = intercept_loc
                print("The model has converged.")
                return {"beta_hat": self.beta_hat, "gamma_hat": self.gamma_hat, "delta_hat": self.delta_hat,
                        "tau_hat": self.tau_hat, "beta_all": beta_all, "intercept": self.intercept}

            rel_dif = compute_rel_dif(Q_old=Q_old, Q_new=Q_new)
            #print("  Relative difference is ", rel_dif)
            #print("Q_new old: ", Q_new, Q_old)

            if abs(rel_dif) <= tol:
                self.beta_hat, self.gamma_hat, self.delta_hat, self.tau_hat = beta_hat, gamma_hat, delta_hat, tau_hat
                self.beta_all = beta_all
                self.intercept = intercept_loc
                print("The model has converged.")
                #print("Last self.intercept", self.intercept, intercept_loc)
                return {"beta_hat": self.beta_hat, "gamma_hat": self.gamma_hat, "delta_hat": self.delta_hat,
                        "tau_hat": self.tau_hat, "beta_all": beta_all, "intercept": self.intercept}

            Q_old = Q_new  # advance baseline

        print("It has not converged. The max number of iterations can be increased.")
        self.beta_hat, self.gamma_hat, self.delta_hat, self.tau_hat = beta_hat, gamma_hat, delta_hat, tau_hat
        self.beta_all = beta_all
        self.intercept = intercept_loc
        return {"beta_hat": self.beta_hat, "gamma_hat": self.gamma_hat, "delta_hat": self.delta_hat,
                "tau_hat": self.tau_hat, "beta_all": beta_all, "intercept": self.intercept}

    def predict(self, X_new):
        beta_all = self.beta_all
        v = X_new @ beta_all + self.intercept
        y_pred = kappa1(v)
        return y_pred

    def R2_score(self, X_new, y_true, verbose=True):
        y_pred = self.predict(X_new)
        if verbose:
            print("R2 score is", r2(y_true, y_pred))
        return r2(y_true, y_pred)
